File: packages/r-blockwise/R/blocks.R
#' Estimate the number of blocks via the elbow heuristic
#'
#' Applies k-means to the binary missingness-indicator matrix for
#' k = 1, ..., \code{k_max} and records, for each k, the fraction of
#' still-missing cells after dropping columns that are sparse within each
#' cluster. The curve is monotone-decreasing; BRM picks the k at its elbow.
#'
#' Typically called internally by \code{\link{brm}} when \code{n_blocks = NULL}.
#'
#' @param X A data.frame of predictors; may contain \code{NA}.
#' @param low_threshold Fraction below which a column is considered absent in
#'   a candidate subset. Default 0.05.
#' @param n_restarts Number of k-means restarts per k. Default 10.
#' @param k_max Upper bound on k. Default \code{min(ncol(X), 50)}.
#' @return A list with \code{n_blocks} (the chosen k) and \code{missing_curve}
#'   (numeric vector of length \code{k_max}).
#' @references Srinivasan, K., Currim, F., Ram, S. (2025). A Reduced Modeling
#'   Approach for Making Predictions With Incomplete Data Having Blockwise
#'   Missing Patterns. \emph{INFORMS Journal on Data Science}.
#' @importFrom stats kmeans
#' @export
choose_num_blocks <- function(X, low_threshold = 0.05, n_restarts = 10, k_max = NULL) {
  if (nrow(X) > 50000L) {
    set.seed(12345)
    X <- X[sample(nrow(X), 50000L, replace = FALSE), , drop = FALSE]
  }
  M <- as.data.frame(
    matrix(as.integer(!is.na(X)), nrow = nrow(X), ncol = ncol(X))
  )
  names(M) <- names(X)

  if (is.null(k_max)) k_max <- min(ncol(X), 50L)
  curve <- vapply(
    seq_len(k_max),
    function(k) subset_missing_frac(M, k, low_threshold, n_restarts),
    numeric(1)
  )
  list(n_blocks = elbow_point(curve), missing_curve = curve)
}

# Best-of-n k-means run for a given k (internal).
best_kmeans <- function(M, k, n_restarts = 10L) {
  if (k == 1L) {
    centers <- matrix(colMeans(M), nrow = 1L)
    colnames(centers) <- colnames(M)
    return(list(cluster = rep(1L, nrow(M)), centers = centers,
                totss = 0, tot.withinss = 0))
  }
  runs <- lapply(seq_len(n_restarts), function(s) {
    set.seed(s)
    stats::kmeans(M, centers = k)
  })
  scores <- vapply(runs,
                   function(r) 100 - r$tot.withinss / r$totss * 100,
                   numeric(1))
  runs[[which.max(scores)]]
}

# Missing-cell density after clustering rows by their missingness pattern.
subset_missing_frac <- function(M, k, low_threshold = 0.05, n_restarts = 10L) {
  km <- best_kmeans(M, k, n_restarts)
  parts <- split(M, km$cluster)
  total_missing <- 0
  for (p in parts) {
    p <- as.data.frame(p)
    completeness <- colSums(p)
    keep <- completeness > low_threshold * nrow(p)
    if (!any(keep)) next
    p_keep <- p[, keep, drop = FALSE]
    total_missing <- total_missing + sum(nrow(p_keep) - colSums(p_keep))
  }
  total_missing / (nrow(M) * ncol(M))
}

# Elbow = point most distant from the line (x1,y1)-(xN,yN).
elbow_point <- function(y) {
  x <- seq_along(y)
  if (length(x) <= 2L) return(1L)
  x1 <- x[1L]; y1 <- y[1L]
  x2 <- x[length(x)]; y2 <- y[length(y)]
  num <- abs((y2 - y1) * x - (x2 - x1) * y + x2 * y1 - y2 * x1)
  den <- sqrt((y2 - y1)^2 + (x2 - x1)^2)
  if (den == 0) return(1L)
  which.max(num / den)
}

File: packages/py-blockwise/src/blockwise/blocks.py
"""Block-count selection heuristic (elbow on the missing-pattern curve)."""
from __future__ import annotations

from typing import Optional

import numpy as np
import pandas as pd
from sklearn.cluster import KMeans


def _best_kmeans(M: np.ndarray, k: int, n_restarts: int = 10) -> KMeans:
    """Best of `n_restarts` k-means fits on the binary missingness matrix."""
    if k == 1:
        km = KMeans(n_clusters=1, n_init=1, random_state=0).fit(M)
        return km
    runs = [
        KMeans(n_clusters=k, n_init=1, random_state=s).fit(M)
        for s in range(n_restarts)
    ]
    # Paper's score: 100 - tot.withinss/totss * 100 — equivalent to smallest inertia.
    return min(runs, key=lambda r: r.inertia_)


def _subset_missing_frac(
    M: np.ndarray, k: int, low_threshold: float, n_restarts: int
) -> float:
    km = _best_kmeans(M, k, n_restarts)
    n, p = M.shape
    total_missing = 0
    for c in np.unique(km.labels_):
        idx = km.labels_ == c
        part = M[idx]
        completeness = part.sum(axis=0)
        keep = completeness > low_threshold * part.shape[0]
        if not keep.any():
            continue
        kept = part[:, keep]
        total_missing += (kept.shape[0] - kept.sum(axis=0)).sum()
    return total_missing / (n * p)


def _elbow_point(y: np.ndarray) -> int:
    """Return 1-indexed k at the elbow (max distance from first-last line)."""
    x = np.arange(1, len(y) + 1)
    if len(x) <= 2:
        return 1
    x1, y1 = x[0], y[0]
    x2, y2 = x[-1], y[-1]
    num = np.abs((y2 - y1) * x - (x2 - x1) * y + x2 * y1 - y2 * x1)
    den = np.sqrt((y2 - y1) ** 2 + (x2 - x1) ** 2)
    if den == 0:
        return 1
    return int(np.argmax(num / den)) + 1


def choose_num_blocks(
    X: pd.DataFrame,
    low_threshold: float = 0.05,
    n_restarts: int = 10,
    k_max: Optional[int] = None,
    random_state: int = 12345,
) -> dict:
    """Elbow heuristic for the number of blocks.

    Parameters
    ----------
    X : pandas.DataFrame
        Predictor frame; may contain ``NaN``.
    low_threshold : float
        Column-density threshold: in a candidate cluster, columns observed
        in fewer than ``low_threshold * n_rows`` rows are treated as absent.
    n_restarts : int
        k-means restarts per candidate k.
    k_max : int or None
        Upper bound on k (default ``min(n_features, 50)``).
    random_state : int
        Seed used when subsampling very large frames.

    Returns
    -------
    dict
        ``{"n_blocks": int, "missing_curve": np.ndarray}``.
    """
    if not isinstance(X, pd.DataFrame):
        X = pd.DataFrame(X)
    if len(X) > 50_000:
        rng = np.random.default_rng(random_state)
        X = X.iloc[rng.choice(len(X), 50_000, replace=False)]
    M = (~X.isna()).to_numpy().astype(np.int8)
    if k_max is None:
        k_max = min(X.shape[1], 50)
    curve = np.array([
        _subset_missing_frac(M, k, low_threshold, n_restarts)
        for k in range(1, k_max + 1)
    ])
    return {"n_blocks": _elbow_point(curve), "missing_curve": curve}
